src/database/transaction.rs

```diff
@@ -1,87 +1,180 @@
+        r#"
                 UPDATE identities
                 SET    status = $2, mined_at = NULL
                 WHERE  id > $1
                 "#,
-            )
-            .bind(root_id)
-            .bind(<&str>::from(ProcessedStatus::Pending));
+    )
+    .bind(root_id)
+    .bind(<&str>::from(ProcessedStatus::Pending));
 
-            tx.execute(update_previous_roots).await?;
-            tx.execute(update_next_roots).await?;
+    tx.execute(update_previous_roots).await?;
+    tx.execute(update_next_roots).await?;
 
-            Ok(())
-        })
-        .await
-    }
+    Ok(())
+}
 
-    /// Marks the identities and roots from before a given root hash as
-    /// finalized
-    #[instrument(skip(self), level = "debug")]
-    pub async fn mark_root_as_mined_tx(&self, root: &Hash) -> Result<(), Error> {
-        let mined_status = ProcessedStatus::Mined;
+pub async fn mark_root_as_mined(
+    tx: &mut Transaction<'_, Postgres>,
+    root: &Hash,
+) -> Result<(), Error> {
+    let mined_status = ProcessedStatus::Mined;
 
-        retry_tx!(self.pool, tx, {
-            let root_id = tx.get_id_by_root(root).await?;
+    let root_id = tx.get_id_by_root(root).await?;
 
-            let Some(root_id) = root_id else {
-                return Err(Error::MissingRoot { root: *root });
-            };
+    let Some(root_id) = root_id else {
+        return Err(Error::MissingRoot { root: *root });
+    };
 
-            let root_id = root_id as i64;
+    let root_id = root_id as i64;
 
-            let update_previous_roots = sqlx::query(
-                r#"
+    let update_previous_roots = sqlx::query(
+        r#"
                 UPDATE identities
                 SET    status = $2
                 WHERE  id <= $1
                 AND    status <> $2
                 "#,
-            )
-            .bind(root_id)
-            .bind(<&str>::from(mined_status));
+    )
+    .bind(root_id)
+    .bind(<&str>::from(mined_status));
+
+    tx.execute(update_previous_roots).await?;
+
+    Ok(())
+}
+
+pub async fn insert_identities_batch(
+    tx: &mut Transaction<'_, Postgres>,
+    latest_tree: &TreeVersion<Latest>,
+    identities: &[UnprocessedCommitment],
+    pending_insertions_mutex: &Mutex<()>,
+) -> anyhow::Result<()> {
+    // Filter out any identities that are already in the `identities` table
+    let mut filtered_identities = vec![];
+    for identity in identities {
+        if tx
+            .get_identity_leaf_index(&identity.commitment)
+            .await?
+            .is_some()
+        {
+            tracing::warn!(?identity.commitment, "Duplicate identity");
+            tx.remove_unprocessed_identity(&identity.commitment).await?;
+        } else {
+            filtered_identities.push(identity.commitment);
+        }
+    }
+
+    let _guard = pending_insertions_mutex.lock().await;
+
+    let next_db_index = tx.get_next_leaf_index().await?;
+    let next_leaf = latest_tree.next_leaf();
```

Comment:
This line makes me unsure about moving whole function to database module. It is not only modifying database but also latest tree mixing a lot of responsibilities in database module now. What do you think?

Reply:
Yeah I'm with you. I was internally struggling about where these things belong. On the one hand, it would be nice to have everything that interacts with the database within the database mod. This may be unrealistic and perhaps the concepts of transactions should be allowed to spill over into other code.